7주차 과제를 제출합니다

Kwakcena opened the pull request:
늦어서 죄송합니다.

Kwakcena committed:
Spring security 설정을 추가하라

Spring security의 기능을 사용하기 위해 build에 Spring security를
추가하고, 이에 대한 config를 구현하였습니다.

Kwakcena committed:
csrf 기능을 끕니다

Security의 csrf 기능을 사용하지 않습니다.

Kwakcena committed:
authentication filter를 추가하라

요청이 들어왔을 때 컨트롤러로 넘어가기 전 실행되는 filter를
정의할 수 있습니다.

Kwakcena committed:
interceptor의 역할을 Security로 옮긴다

Security 로 interceptor가 하는 역할을 옮겼습니다.
요청에 따라 accessToken이 필요한지 불필요한지 구분하고,
이에 따라 권한을 부여하도록 합니다.

Kwakcena committed:
JwtAuthenticationFilter 실행 전에 Error advice를 실행하라

JwtAuthenticationFilter가 실행되기 전에 AuthenticationErrorFilter를 동작시켜서
에러가 발생할 경우 WebMvc에서 advice를 해 준 것처럼 동작시킬 수 있다. 이에 대하여
구현하고 실행하도록 합니다.

Kwakcena committed:
PreAuthorize를 등록하여 인증 절차를 거친다

컨트롤러 요청에 PreAuthorize를 사용하여 인증된 사용자인지
먼저 확인하는 절차를 구현한다.

Kwakcena committed:
sessionManagement를 등록한다

sessionManagement를 등록하여 state less 한 rest full api 요청에 맞도록 설정한다

Kwakcena committed:
user의 비밀번호를 암호화 해서 저장한다

user의 비밀번호를 암호화 하여 저장하도록 합니다.

Kwakcena committed:
암호화 된 비밀번호가 평문과 같은지 비교한다

암호화 된 비밀번호가 평문과 같은지 비교하는 로직을 추가한다

Kwakcena committed:
User가 어떤 방식으로 비밀번호를 encode하는 지 모르게 한다

의존성 주입을 통하여 User 도메인에 passwordEncoder 클래스를 주입시키도록 하여
User 도메인 내에서 passwordEncoder에 대해 모르게 한다.

Kwakcena committed:
Apply forbidden exception

user를 업데이트 할 때 다른 id로 요청할 경우 금지된 요청에 대한
예외를 응답해야 한다.

hannut91 commented on app/src/main/java/com/codesoom/assignment/config/SecurityJavaConfig.java:
```diff
@@ -0,0 +1,38 @@
+        Filter authenticationFilter = new JwtAuthenticationFilter(authenticationManager(), authenticationService);
+        Filter authenticationErrorFilter = new AuthenticationErrorFilter();
+
+        http
+                .csrf().disable()
+                .addFilter(authenticationFilter)
+                .addFilterBefore(authenticationErrorFilter, JwtAuthenticationFilter.class)
+                .sessionManagement()
+                    .sessionCreationPolicy(SessionCreationPolicy.STATELESS)
+                .and()
+                .exceptionHandling()
+                .authenticationEntryPoint(new HttpStatusEntryPoint(HttpStatus.UNAUTHORIZED));
```
Spring security를 사용하면, 우리가 직접 구현했던 많은 부분들이 마법처럼 만들어지죠.
우리가 주어진 시간안에 학습하며 일을 하다보면 동작하는 것만 확인하고 넘어가야 할 때가 많습니다.
나중에 시간을 내서 더 자세하게 왜 이렇게 동작하는건지 학습해야 합니다.
이미 동작하는 것을 알고 있으니, 왜 이렇게 동작하는지 학습하는 것은 어렵지 않을 거에요.(물론 스프링 시큐리티가 좀 어렵긴 하니, 그 나마 낫다고 표현해야 겠네요)

See also
  - https://docs.spring.io/spring-security/reference/index.html

hannut91 commented on app/src/main/java/com/codesoom/assignment/security/UserAuthentication.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public String toString() {
+        return "UserAuthentication{" +
+                "userId=" + userId +
+                '}';
```
`String.format`을 사용해 보세요

Kwakcena committed:
test를 위해 in-memory를 사용합니다

in-memory를 사용하여 테스트 합니다.

Kwakcena committed:
전체 유저를 삭제하는 매소드를 추가합니다

deleteAll을 추가하였습니다.

Kwakcena committed:
회원가입 테스트 리팩토링

회원가입 테스트를 리팩토링 하였습니다.

Kwakcena committed:
userId를 encode 한 accessToken을 반환한다

jwtUtil의 encode 메소드가 하드코딩된 userId를 encode하는데
이를 입력받은 매개변수를 encode 하도록 수정하였습니다.

Kwakcena committed:
TestUtils 를 만들어라

TestUtils 은 Test에서 사용할 util 입니다.
어떤 요청의 결과를 TestUtils의 content 메소드에 매개변수로
넣어주면, 해당 요청의 응답값을 꺼내 objectMapper를 이용하여
type 매개변수의 클래스에 맞는 java object로 바꿔줍니다.

Kwakcena committed:
AllArgsConstructor를 제거하라

AllArgsConstructor 어노테이션을 제거합니다.

Kwakcena committed:
UserController update user 리팩토링

리팩토링을 진행중입니다.

hannut91 commented on app/src/test/java/com/codesoom/assignment/TestUtils.java:
```diff
@@ -0,0 +1,20 @@
+
+import java.io.UnsupportedEncodingException;
+
+public class TestUtils {
```
이건 어디에 쓰이나요?
